AwsCryptographyPrimitives/src/HKDF/HMAC.dfy

```diff
@@ -56,8 +56,20 @@ module {:options "-functionSyntax:4"} {:extern "HMAC"} HMAC {
       ensures this.GetKey() == old(this.GetKey())
       ensures this.HashSignature(old(this.GetInputSoFar()), s)
 
-    predicate {:axiom} HashSignature(message: seq<uint8>, s: seq<uint8>)
+    ghost predicate {:axiom} HashSignature(message: seq<uint8>, s: seq<uint8>)
 
+    // The next two functions are for the benefit of the extern implementation to call,
+    // avoiding direct references to generic datatype constructors
+    // since their calling pattern is different between different versions of Dafny
+    // (i.e. after 4.2, explicit type descriptors are required).
+
+    static function CreateHMacSuccess(hmac: HMac): Result<HMac, Types.Error> {
+      Success(hmac)
+    }
+
+    static function CreateHMacFailure(error: Types.Error): Result<HMac, Types.Error> {
+      Failure(error)
+    }
```

Comment:
Unfortunately no, and we had the same conversation over the smithy-dafny change: https://github.com/smithy-lang/smithy-dafny/pull/301#issuecomment-1788072993

The best comment explaining it is here: https://github.com/aws/aws-cryptographic-material-providers-library/pull/195/files#diff-cde85dd0ffecd09462b9d1bec99964d0aef44bb2413c43f3f95c213339ac932fR6-R25. I'm open to ideas about how to communicate this more thoroughly though as I'm sure it will keep coming up. :)